xlators/cluster/afr/src/afr-open.c

```diff
@@ -271,34 +295,192 @@ afr_fd_ctx_need_open(fd_t *fd, xlator_t *this, unsigned char *need_open)
+    int i = 0;
+    int need_open_count = 0;
+
+    priv = this->private;
+
+    need_open_count = AFR_COUNT(local->need_open, priv->child_count);
+    if (!need_open_count) {
+        goto out;
+    }
+    gf_msg_debug(this->name, 0, "need open count: %d", need_open_count);
+    local->call_count = need_open_count;
+
+    for (i = 0; i < priv->child_count; i++) {
+        if (!local->need_open[i])
+            continue;
+
+        if (IA_IFDIR == local->fd->inode->ia_type) {
+            gf_msg_debug(this->name, 0, "opening fd for dir %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+            STACK_WIND_COOKIE(frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                              priv->children[i],
+                              priv->children[i]->fops->opendir, &local->loc,
+                              local->fd, NULL);
+        } else {
+            gf_msg_debug(this->name, 0,
+                         "opening fd for file %s on subvolume %s",
+                         local->loc.path, priv->children[i]->name);
+
+            STACK_WIND_COOKIE(
+                frame, afr_openfd_fix_open_cbk, (void *)(long)i,
+                priv->children[i], priv->children[i]->fops->open, &local->loc,
+                local->fd_ctx->flags & ~(O_CREAT | O_EXCL | O_TRUNC), local->fd,
+                NULL);
+        }
+        if (!--need_open_count)
+            break;
+    }
+    return 0;
+
+out:
+    AFR_STACK_DESTROY(frame);
+    return 0;
+}
+
+static int
+afr_is_reopen_allowed_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
+                          int32_t op_ret, int32_t op_errno,
+                          struct gf_flock *lock, dict_t *xdata)
+{
+    afr_local_t *local = frame->local;
+    afr_private_t *priv = NULL;
+    unsigned char *success_replies = NULL;
+    int ret = -1;
+    int call_count = 0;
+    int i = (long)cookie;
+
+    priv = this->private;
+    local->replies[i].valid = 1;
+    local->replies[i].op_ret = op_ret;
+    local->replies[i].op_errno = op_errno;
+    if (op_ret != 0) {
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, AFR_MSG_LK_HEAL_DOM,
+               "Failed getlk for %s", uuid_utoa(local->fd->inode->gfid));
+    }
+
+    LOCK(&frame->lock);
+    {
+        if (local->is_reopen_allowed) {
+            ret = dict_get_int32(xdata, "is-reopen-allowed-on-fd",
+                                 &local->is_reopen_allowed);
+            if (ret)
+                gf_msg(this->name, GF_LOG_ERROR, -ret, AFR_MSG_DICT_GET_FAILED,
+                       "Failed to get whether reopen is allowed or not on FD.");
+        }
+    }
+    UNLOCK(&frame->lock);
+
+    call_count = afr_frame_return(frame);
+
+    if (call_count)
+        return 0;
+
+    success_replies = alloca0(priv->child_count);
+    afr_fill_success_replies(local, priv, success_replies);
+    if (AFR_COUNT(success_replies, priv->child_count) == 0 &&
+        !local->is_reopen_allowed) {
```

Comment:
If the afr_is_reopen_allowed_cbk() gets op_ret=-1 for all bricks, then success_replies will be zero and local->is_reopen_allowed will be 1 (init value).  Shouldn't we fail the open in that case?

Reply:
Yes, we should fail in that scenario as well.
Removed this success_replies check and getting different values from the lower layer (protocol/client in this case)  now to handle different scenarios. Done this in such a way that, if a new layer is added in between in later point in time which can also decide on the reopen, and only wants to send "is_reopen_allowed" value set to either 0 or 1, afr can handle that case as well, without any changing the current code.